src/generic_solver.cpp

```diff
@@ -478,31 +481,121 @@ Sort GenericSolver::make_sort(SortKind sk, const SortVec & sorts) const
+
+      to_solver += ")";
+    }
+    to_solver += ")\n))";
+    assert(name_sort_map->find(dt_decl_name) == name_sort_map->end());
+    (*name_sort_map)[dt_decl_name] = dt_sort;
+    (*sort_name_map)[dt_sort] = dt_decl_name;
+    run_command(to_solver);
+    
+    return dt_sort;
+  }
+  else
+  {
+    throw IncorrectUsageException(string("sort name: ") + dt_decl_name
+                                  + string(" already taken"));
+  }
 }
-  
+
 DatatypeDecl GenericSolver::make_datatype_decl(const std::string & s)
 {
-  throw NotImplementedException("Generic Solvers do not support datatypes");
+  DatatypeDecl new_dt_decl = make_shared<GenericDatatypeDecl>(s);
+  shared_ptr<GenericDatatype> new_dt =
+      shared_ptr<GenericDatatype>(new GenericDatatype(new_dt_decl));
+  (*name_datatype_map)[s] = new_dt;
+  (*datatype_name_map)[new_dt] = s;
+  return new_dt_decl;
 }
 DatatypeConstructorDecl GenericSolver::make_datatype_constructor_decl(
     const std::string s)
 {
-  throw NotImplementedException("Generic Solvers do not support datatypes");
+  shared_ptr<GenericDatatypeConstructorDecl> new_dt_cons_decl =
+      shared_ptr<GenericDatatypeConstructorDecl>(
+          new GenericDatatypeConstructorDecl(s));
+  return new_dt_cons_decl;
 }
-  
+
 void GenericSolver::add_constructor(DatatypeDecl & dt, const DatatypeConstructorDecl & con) const
   {
-    throw NotImplementedException("Generic Solvers do not support datatypes");
+    shared_ptr<GenericDatatypeDecl> gdt_decl =
+        static_pointer_cast<GenericDatatypeDecl>(dt);
+    string name = gdt_decl->get_name();
+    auto gdt = (*name_datatype_map)[name];
+    gdt->add_constructor(con);
 }
+
 void GenericSolver::add_selector(DatatypeConstructorDecl & dt, const std::string & name, const Sort & s) const
 {
-  throw NotImplementedException("Generic Solvers do not support datatypes");
+  shared_ptr<SelectorComponents> newSelector =
+      make_shared<SelectorComponents>();
+  newSelector->name = name;
+  newSelector->sort = s;
+  newSelector->finalized = true;
+  shared_ptr<GenericDatatypeConstructorDecl> gdtc =
+      static_pointer_cast<GenericDatatypeConstructorDecl>(dt);
+  gdtc->add_new_selector(*newSelector);
 }
   
 void GenericSolver::add_selector_self(DatatypeConstructorDecl & dt, const std::string & name) const
   {
-    throw NotImplementedException("Generic Solvers do not support datatypes");
+    shared_ptr<SelectorComponents> newSelector =
+        make_shared<SelectorComponents>();
+    shared_ptr<GenericDatatypeConstructorDecl> gdt_cons = static_pointer_cast<GenericDatatypeConstructorDecl>(dt);
+    string dt_decl_name = gdt_cons->get_dt_name();
+    
+    newSelector->name = name;
+    newSelector->sort = make_shared<GenericSort>(name);
+    newSelector->finalized = false;
+    assert(name_datatype_map->find(dt_decl_name) != name_datatype_map->end());
+    shared_ptr<GenericDatatype> curr_dt = (*name_datatype_map)[dt_decl_name];
+    gdt_cons->add_new_selector(*newSelector);
```

Comment:
Please add documentation explaining your code, generally, but here in particular because we are doing something special here with the placeholder sort.